### DISCUSSION TITLE
v4.0.0-b0

### DISCUSSION THREAD

**Discussion by not522** [optuna/optuna#5585] [Announcements]
This is the release note of [v4.0.0-b0](https://github.com/optuna/optuna/milestone/61?closed=1).

If you want to update your existing projects from Optuna v3.x to Optuna v4, please see the [migration guide](https://github.com/optuna/optuna/discussions/5573) and try out Optuna v4.

# Highlights

## OptunaHub Beta Release

The Optuna team released the beta version of OptunaHub, the feature-sharing platform for Optuna. Registered features can be easily implemented on users’ code and contributors can register the features they implement. The beta version of OptunaHub is now ready to accept contributions from all over the world. Visit [hub.optuna.org](https://hub.optuna.org)!

The following code shows an example to use a sampler registered on OptunaHub.

```bash
% pip install optunahub
```

```python
import optunahub
import optuna

def objective(trial):
    x = trial.suggest_float("x", 0, 1)
    return x

mod = optunahub.load_module("samplers/simulated_annealing")

sampler = mod.SimulatedAnnealingSampler()
study = optuna.create_study(sampler=sampler)
study.optimize(objective, n_trials=20)
```

## Stabilization of Artifact

The stable version of the artifact module is available, now equipped with several new APIs. This module introduces capabilities for managing the relatively large-sized data such as model snapshots in hyperparameter tuning, training/validation datasets, and etc. Compared to third-party libraries for experiment tracking, the advantage of using Optuna’s artifact module is a tight integration of [Optuna Dashboard](https://github.com/optuna/optuna-dashboard). This allows users to see artifacts (files) associated with the Optuna trial or study.

Here is a list of new APIs:
- [`download_artifact`](https://optuna.readthedocs.io/en/v4.0.0-b0/reference/artifacts.html#optuna.artifacts.download_artifact): Download an artifact from the artifact store.
- [`get_all_artifact_meta`](https://optuna.readthedocs.io/en/v4.0.0-b0/reference/artifacts.html#optuna.artifacts.get_all_artifact_meta): List the associated artifact information of the provided trial or study.

## Stabilization of `JournalStorage`

The stable version of `JournalStorage` is available. This implies we have decided to maintain backward compatibility of the log format in the future releases.

Please note that this release introduces the following API changes to improve the clarity of class names and the module structure.

| Deprecated APIs | Corresponding active APIs |
-|-
| `optuna.storages.JournalFileStorage` | `optuna.storages.journal.JournalFileBackend` |
| `optuna.storages.JournalFileSymlinkLock` | `optuna.storages.journal.JournalFileSymlinkLock` |
| `optuna.storages.JournalFileOpenLock` | `optuna.storages.journal.JournalFileOpenLock` |
| `optuna.storages.JournalRedisStorage` | `optuna.storages.journal.JournalRedisBackend` |

# Breaking Changes

- Delete deprecated three integrations, `skopt`, `catalyst`, and `fastaiv1` (https://github.com/optuna/optuna-integration/pull/114)
- Remove deprecated `CmaEsSampler` from integration (https://github.com/optuna/optuna-integration/pull/116)
- Remove verbosity of `LightGBMTuner` (https://github.com/optuna/optuna-integration/pull/136)
- Move positional args of LightGBM tuner (https://github.com/optuna/optuna-integration/pull/138)
- Remove `multi_objective` (#5390)
- Delete deprecated `_ask` and `_tell` (#5398)
- Delete deprecated `--direction(s)` arguments in the `ask` command (#5405)
- Delete deprecated three integrations, `skopt`, `catalyst`, and `fastaiv1` (#5407)
- Remove the default normalization of importance in f-ANOVA (#5411)
- Remove `samplers.intersection` (#5414)
- Drop implicit create-study in `ask` command (#5415)
- Remove deprecated `study optimize` CLI command (#5416)
- Remove deprecated `CmaEsSampler` from integration (#5417)
- Support constrained optimization in `best_trial` (#5426)
- Drop `--study` in `cli.py` (#5430)
- Deprecate `constraints_func` in `plot_pareto_front` function (#5455)
- Rename some classnames related to `JournalStorage` (#5539)

# New Features

- Add Comet ML integration (https://github.com/optuna/optuna-integration/pull/63, thanks @caleb-kaiser!)
- Add Knowledge Gradient candidates functions (https://github.com/optuna/optuna-integration/pull/125, thanks @alxhslm!)
- Add `is_exhausted()` function in the `GridSampler` class (#5306, thanks @aaravm!)
- Remove experimental from plot (#5413)
- Implement `download_artifact` (#5448)
- Add a function to list linked artifact information (#5467)
- Stabilize artifact APIs (#5567)
- Stabilize `JournalStorage` (#5568)

# Enhancements

- Pass two arguments to the forward of `ConstrainedMCObjective` to support `botorch=0.10.0` (https://github.com/optuna/optuna-integration/pull/106)
- Speed up non-dominated sort (#5302)
- Make 2d hypervolume computation twice faster (#5303)
- Reduce the time complexity of HSSP 2d from `O(NK^2 log K)` to `O((N - K)K)` (#5346)
- Introduce lazy hypervolume calculations in HSSP for speedup (#5355)
- Make `plot_contour` faster (#5369)
- Speed up `to_internal_repr` in `CategoricalDistribution` (#5400)
- Allow users to modify categorical distance more easily (#5404)
- Speed up `WFG` by NumPy vectorization (#5424)
- Check whether the study is multi-objective in `sample_independent` of `GPSampler` (#5428)
- Suppress warnings from `numpy` in hypervolume computation (#5432)
- Make an option to assume Pareto optimality in WFG (#5433)
- Adapt multi objective to NumPy v2.0.0 (#5493)
- Enhance the error message for integration installation (#5498)
- Reduce journal size of `JournalStorage` (#5526)
- Simplify a SQL query for getting the `trial_id` of `best_trial` (#5537)

# Bug Fixes

- Log `None` objective trials correctly (https://github.com/optuna/optuna-integration/pull/119, thanks @neel04!)
- Pass two arguments to the forward of `ConstrainedMCObjective` in `qnei_candidates_func` (https://github.com/optuna/optuna-integration/pull/124, thanks @alxhslm!)
- Allow single split cv in `OptunaSearchCV` (https://github.com/optuna/optuna-integration/pull/128, thanks @sgerloff!)
- Fix `WilcoxonPruner` bug when `best_trial` has no intermediate value (#5354)
- Debug an error caused by convergence in `GPSampler` (#5359)
- Fix `average_is_best` implementation in `WilcoxonPruner` (#5366)
- Create an unique renaming filename for each release operation in lock systems of `JournalStorage` (#5389)
- Fix `_normalize_value` for incomplete trials (#5422)
- Fix heartbeat for race condition (#5431)
- Guarantee `weights_below` to be finite in MOTPE (#5435)
- Fix `_log_complete_trial` for constrained optimization (#5462)
- Convert `step` to `int` in `report` (#5488)
- Use a fixed seed value when `seed=None` in `GridSampler` (#5490)
- Acquire session lock only for write in heartbeat (#5496)
- Fix a bug in `_create_new_trial` and refactor it (#5497)
- Add rdb create new trial test (#5525)
- Fix a bug in `sample_normalized_param` for `GPSampler` (#5543)

# Installation

- Add an option to install integration dependencies via pip (https://github.com/optuna/optuna-integration/pull/130)
- Add version constraint to numpy (https://github.com/optuna/optuna-integration/pull/131)

# Documentation

- Enhance `README.md` (https://github.com/optuna/optuna-integration/pull/126)
- Add document page and fix docstring and comments (https://github.com/optuna/optuna-integration/pull/134)
- Use sphinx rst syntax (#5345)
- Fix typo (#5351)
- Update docs about `show_progress_bar` (#5393)
- Fix artifact tutorial (#5451)
- Revise docs to specify `ArtifactStore` methods as non-public (#5474)
- Improve the docstring of `JournalFileStorage` (#5475)
- Improve `make clean` in `docs` (#5487)
- Improve the example of chemical structures in `optuna artifact tutorial` (#5491)
- Add FAQ for artifact store remove API (#5501)
- Add a documentation for `ArtifactMeta` (#5511)
- Change docs version from latest to stable (#5518)
- Add a list of supported storage for artifact store (#5532)
- Rename filenames for `JournalFileStorage` in  documents and a tutorial (#5535)
- Add explanation for `lock_obj` to `JournalFileStorage` docstring (#5540)
- Fix storages document sections (#5553)
- Add `Returns` to the docstring of `load_study` (#5554, thanks @kAIto47802!)
- Fixing paths related to `storages.journal` (#5560)
- Rename filename for `JournalFileBackend` in examples (#5562)

# Examples

- Support tensorflow 2.16.1 and separate CI run for tensorflow estimator (https://github.com/optuna/optuna-examples/pull/248)
- Replace deprecated jax function (https://github.com/optuna/optuna-examples/pull/250)
- Drop Python 3.7 support for wandb (https://github.com/optuna/optuna-examples/pull/252)
- Support python 3.12 in the CIs (https://github.com/optuna/optuna-examples/pull/253)
- Remove dask version constraint (https://github.com/optuna/optuna-examples/pull/254)
- Update GitHub actions versions to `actions/checkout@v4` and `actions/setup-python@v5` (https://github.com/optuna/optuna-examples/pull/255)
- Delete an example using deprecated `fastaiv1` (https://github.com/optuna/optuna-examples/pull/256)
- Rename fastai v2 to fastai (https://github.com/optuna/optuna-examples/pull/257)
- Fix CI (https://github.com/optuna/optuna-examples/pull/258)
- Fix path in pruners workflow (https://github.com/optuna/optuna-examples/pull/259)
- Install `tensorflow-cpu` in CIs (https://github.com/optuna/optuna-examples/pull/260)
- Add python 3.12 and run terminator search cv (https://github.com/optuna/optuna-examples/pull/264)
- Organize the directory structure (https://github.com/optuna/optuna-examples/pull/266)
- Fix CI with version constraints of packages (https://github.com/optuna/optuna-examples/pull/267)
- Install `numpy<2.0.0` for catboot example (https://github.com/optuna/optuna-examples/pull/268)

# Tests

- Suppress `ExperimentalWarning`s (https://github.com/optuna/optuna-integration/pull/108)
- Add a unit test for convergence of acquisition function in `GPSampler` (#5365)
- Remove unused block (#5368)
- Implement backward compatibility tests for `JournalStorage` (#5486)

# Code Fixes

- Align run name strategy between MLflowCallback and track_in_mlflow method (https://github.com/optuna/optuna-integration/pull/111, thanks @TTRh!)
- Use `TYPE_CHECKING` for `ObjectiveFuncType` (https://github.com/optuna/optuna-integration/pull/113)
- Ignore `ExperimentalWarning` for `track_in_wandb` (https://github.com/optuna/optuna-integration/pull/122)
- Remove unused module (https://github.com/optuna/optuna-integration/pull/127)
- Apply formatter and follow the Optuna conventions (https://github.com/optuna/optuna-integration/pull/133)
- Fix future annotations in `percentile.py` (#5322, thanks @aaravm!)
- Delete examples directory (#5329)
- Simplify annotations in `optuna/pruners/_hyperband.py` (#5338, thanks @keita-sa!)
- Simplify annotations in `optuna/pruners/_successive_halving.py` and `optuna/pruners/_threshold.py` (#5343, thanks @keita-sa!)
- Reduce test warnings (#5344)
- Simplify type annotations for `storages/_base.py` (#5352, thanks @Obliquedbishop!)
- Use `TYPE_CHECKING` for `Study` in `samplers` (#5391)
- Refactor `_normalize_objective_values` in NSGA-III to supress `RuntimeWarning` (#5399)
- Make `ObjectiveFuncType` available externally (#5401)
- Replace `numpy` with `np` (#5412)
- Bundle experimental feature warning (#5434)
- Simplify annotations in `_brute_force.py` (#5441)
- Add `__future__.annotations` to base sampler (#5442)
- Introduce `__future__.annotations` to TPE-related modules (#5443)
- Adapt to `__future__.annotations` in `optuna/storages/_rdb/models.py` (#5452, thanks @aisha-partha!)
- Debug an unintended assertion error in `GPSampler` (#5484)
- Make `WFG` a function (#5504)
- Enhance the comments in `create_new_trial` (#5510)
- Replace single trailing underscore with double trailing underscore in `.rst` files (#5514, thanks @47aamir!)
- Fix hyperlinks to use double trailing underscores (#5515, thanks @virendrapatil24!)
- Replace `_` with `__` in the link in Sphinx (#5517)
- Refactor `plot_parallel_coordinate()` (#5527, thanks @karthikkurella!)
- Remove an obsolete TODO comment (#5528)
- Unite the argument order of artifact APIs (#5533)
- Make `assume_unique_lexsorted` in `is_pareto_front` required (#5534)
- Simplify annotations in `terminator` module (#5536)
- Simplify the type annotations in `samplers/_qmc.py` (#5538, thanks @kAIto47802!)
- Rename `solution_set` to `loss_vals` in hypervolume computation (#5541)
- Expand the type of `callbacks` in optimize to `Iterable` (#5542, thanks @kz4killua!)
- Organize the directory structure of `JournalStorage` (#5544)
- Add followup of journal storage module structure organization PR (#5546)
- Move `RetryFailedTrialCallback` to `optuna.storages._callbacks` (#5551)
- Change `removed_version` of deprecated `JournalStorage` classes (#5552)
- Reorganize a journal storage module structure (#5555)
- Remove unused private deprecated  class `BaseJournalLogSnapshot` (#5561)

# Continuous Integration

- Split ci (https://github.com/optuna/optuna-integration/pull/98)
- Install `pipdeptree` v2.16.2 to hotfix parse error (https://github.com/optuna/optuna-integration/pull/109)
- Remove dask version constraint (https://github.com/optuna/optuna-integration/pull/112)
- Rename CI name (https://github.com/optuna/optuna-integration/pull/120)
- Add `deprecated` arg for the comet CI job (https://github.com/optuna/optuna-integration/pull/135)
- Rename CI jobs (#5353, thanks @Obliquedbishop!)
- Update GitHub actions versions of `actions/setup-python` and `actions/checkout` (#5367)
- Add `type: ignore` for CI hotfix (#5419)
- Use CPU-only PyTorch wheels on GitHub Actions (#5465)
- Add a version constraint to NumPy (#5492)
- Add the note of the Hotfix in the workflow file (#5494)
- Remove unnecessary version constraint from workflow (#5495)
- Remove skipped tests for BoTorch with Python 3.12 (#5519)
- Fix the version conflicts in the workflow for the minimum version constraints (#5523)
- Replace `sleep` function with spin waiting to stabilize the frequent Mac test failure (#5549)

# Other

- Bump up version number to 4.0.0dev (https://github.com/optuna/optuna-integration/pull/102)
- Update PyPI classifiers for Python 3.11 support (https://github.com/optuna/optuna-integration/pull/129)
- Bump up version number to 4.0.0b0 (https://github.com/optuna/optuna-integration/pull/139)
- Bump up to version number v4.0.0.dev (#5319)
- Bump up to version number 4.0.0b0 (#5572)

# Thanks to All the Contributors!

This release was made possible by the authors and the people who participated in the reviews and discussions.

@47aamir, @Alnusjaponica, @HideakiImamura, @Obliquedbishop, @TTRh, @aaravm, @aisha-partha, @alxhslm, @c-bata, @caleb-kaiser, @contramundum53, @eukaryo, @gen740, @kAIto47802, @karthikkurella, @keisuke-umezawa, @keita-sa, @kz4killua, @nabenabe0928, @neel04, @not522, @nzw0301, @porink0424, @sgerloff, @toshihikoyanase, @virendrapatil24, @y0z

<hr /><em>This discussion was created from the release <a href='https://github.com/optuna/optuna/releases/tag/v4.0.0-b0'>v4.0.0-b0</a>.</em>
